gusto/transport_equation.py

```diff
@@ -311,28 +311,37 @@ class VectorInvariant(TransportEquation):
     :arg solver_params: (optional) dictionary of solver parameters to pass to the
                         linear solver.
     """
-    def __init__(self, state, V, ibp="once", solver_params=None):
-        super(VectorInvariant, self).__init__(state, V, ibp, solver_params)
+    def __init__(self, physical_domain, V, Vu, ibp="once", solver_params=None):
+        super(VectorInvariant, self).__init__(physical_domain, V, Vu, ibp, solver_params)
 
-        self.Upwind = 0.5*(sign(dot(self.ubar, self.n))+1)
+        Upwind = 0.5*(sign(dot(self.ubar, self.n))+1)
 
-        if self.state.mesh.topological_dimension() == 2:
-            self.perp = state.perp
-            if V.extruded:
-                self.perp_u_upwind = lambda q: self.Upwind('+')*state.perp(q('+')) + self.Upwind('-')*state.perp(q('-'))
-            else:
-                outward_normals = CellNormal(state.mesh)
-                self.perp_u_upwind = lambda q: self.Upwind('+')*cross(outward_normals('+'), q('+')) + self.Upwind('-')*cross(outward_normals('-'), q('-'))
-            self.gradperp = lambda u: state.perp(grad(u))
-        elif self.state.mesh.topological_dimension() == 3:
+        if hasattr(physical_domain, "perp"):
+            perp = physical_domain.perp
+            self.perp = perp
+
+        if physical_domain.is_3d:
+            self.is_3d = True
+            self.Upwind = Upwind
             if self.ibp == "twice":
                 raise NotImplementedError("ibp=twice is not implemented for 3d problems")
         else:
-            raise RuntimeError("topological mesh dimension must be 2 or 3")
+            self.is_3d = False
+            if physical_domain.is_extruded:
+                self.perp_u_upwind = (
+                    lambda q: Upwind('+')*perp(q('+')) + Upwind('-')*perp(q('-'))
+                )
+            elif physical_domain.on_sphere:
+                outward_normals = CellNormal(physical_domain.mesh)
+                self.perp_u_upwind = (
+                    lambda q: Upwind('+')*cross(outward_normals('+'), q('+'))
+                    + Upwind('-')*cross(outward_normals('-'), q('-'))
+                )
+            self.gradperp = lambda u: perp(grad(u))
```

Comment:
I sort of don't like setting all these properties on the class in `__init__` that are basically just used as local variables in one other function.  The reason is that it adds to the cognitive load when reading and understanding the code.  I have to worry, is `self.gradperp` required at a later point many times (for example)?  In this case, it turns out no, I think that `gradperp`, `perp` and `perp_u_upwind` and `Upwind` are basically only used in the `advection_term` method.  So why not make them local variables there?

Reply:
ok, yes, that sounds nicer